{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "707880d2-0c75-4839-bf98-365c4d394228",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/matthewbaxter/Documents/obds_course_exrecises/Week_2\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "de98975f-be4f-4e3e-9bf0-5dda31449058",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2e39fce1-f883-4f90-9bd4-a0ee3f356dca",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/ln/b27g5h050b7c4j9sh03qk90c0000gr/T/ipykernel_96485/2357806593.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  gtf = pd.read_csv('Homo_sapiens.GRCh38.107.gtf.gz', sep='\\t', header = None, comment='#')\n"
     ]
    }
   ],
   "source": [
    "gtf = pd.read_csv('Homo_sapiens.GRCh38.107.gtf.gz', sep='\\t', header = None, comment='#')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "05425715-4569-471d-b85d-a0057cce03ae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1471887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   0               1            2        3        4  5  6  7  \\\n",
       "0  1  ensembl_havana         gene  1471765  1497848  .  +  .   \n",
       "1  1  ensembl_havana   transcript  1471765  1497848  .  +  .   \n",
       "2  1  ensembl_havana         exon  1471765  1472089  .  +  .   \n",
       "3  1  ensembl_havana          CDS  1471885  1472089  .  +  0   \n",
       "4  1  ensembl_havana  start_codon  1471885  1471887  .  +  0   \n",
       "\n",
       "                                                   8  \n",
       "0  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "1  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "2  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "3  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "4  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "edba2f13-6e35-4046-b07d-98720d777a40",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3371244, 9)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca00265e-0b73-44b7-8aec-843a404f385b",
   "metadata": {},
   "source": [
    "# Assign column names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "58b59a7c-6167-4ce1-a97a-66c303cdb07f",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf.columns=['seqname', 'source' , 'feature' , 'start' , 'end', 'score' , 'strand' ,  'frame' ,  'attribute' ]\n",
    "                          "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2e7ddaab-fe39-4163-81aa-50e5af47e17b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1471887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname          source      feature    start      end score strand frame  \\\n",
       "0       1  ensembl_havana         gene  1471765  1497848     .      +     .   \n",
       "1       1  ensembl_havana   transcript  1471765  1497848     .      +     .   \n",
       "2       1  ensembl_havana         exon  1471765  1472089     .      +     .   \n",
       "3       1  ensembl_havana          CDS  1471885  1472089     .      +     0   \n",
       "4       1  ensembl_havana  start_codon  1471885  1471887     .      +     0   \n",
       "\n",
       "                                           attribute  \n",
       "0  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "1  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "2  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "3  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "4  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "470b6f19-ff7d-405f-a713-b31e093318ac",
   "metadata": {},
   "source": [
    "# How many unique values for ‘source’ column?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "2fccfb6d-4b6b-41d8-a212-90d5bd4c321d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 3371244 entries, 0 to 3371243\n",
      "Data columns (total 9 columns):\n",
      " #   Column     Dtype \n",
      "---  ------     ----- \n",
      " 0   seqname    object\n",
      " 1   source     object\n",
      " 2   feature    object\n",
      " 3   start      int64 \n",
      " 4   end        int64 \n",
      " 5   score      object\n",
      " 6   strand     object\n",
      " 7   frame      object\n",
      " 8   attribute  object\n",
      "dtypes: int64(2), object(7)\n",
      "memory usage: 231.5+ MB\n"
     ]
    }
   ],
   "source": [
    "gtf.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "139b2ce6-c269-4819-95c4-6d7ede562ff7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['ensembl_havana', 'havana', 'ensembl', 'havana_tagene',\n",
       "       'ensembl_havana_tagene', 'mirbase', 'ensembl_tagene', 'insdc'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf[\"source\"].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a34942e6-4589-429f-b44d-3d5edce0a0e9",
   "metadata": {},
   "source": [
    "# Subset all values from ‘ensembl_havana’ and assign to new df called ‘ens_df’ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "6258c1df-a59f-4375-b57c-67862bf74fb9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1471887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3370044</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>46445284</td>\n",
       "      <td>46445769</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160299\"; gene_version \"19\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3370045</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>46445284</td>\n",
       "      <td>46445324</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"ENSG00000160299\"; gene_version \"19\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3370046</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>46445325</td>\n",
       "      <td>46445327</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160299\"; gene_version \"19\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3370047</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>five_prime_utr</td>\n",
       "      <td>46324156</td>\n",
       "      <td>46324228</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160299\"; gene_version \"19\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3370048</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>46445328</td>\n",
       "      <td>46445769</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160299\"; gene_version \"19\"; ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>714883 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname          source          feature     start       end score  \\\n",
       "0             1  ensembl_havana             gene   1471765   1497848     .   \n",
       "1             1  ensembl_havana       transcript   1471765   1497848     .   \n",
       "2             1  ensembl_havana             exon   1471765   1472089     .   \n",
       "3             1  ensembl_havana              CDS   1471885   1472089     .   \n",
       "4             1  ensembl_havana      start_codon   1471885   1471887     .   \n",
       "...         ...             ...              ...       ...       ...   ...   \n",
       "3370044      21  ensembl_havana             exon  46445284  46445769     .   \n",
       "3370045      21  ensembl_havana              CDS  46445284  46445324     .   \n",
       "3370046      21  ensembl_havana       stop_codon  46445325  46445327     .   \n",
       "3370047      21  ensembl_havana   five_prime_utr  46324156  46324228     .   \n",
       "3370048      21  ensembl_havana  three_prime_utr  46445328  46445769     .   \n",
       "\n",
       "        strand frame                                          attribute  \n",
       "0            +     .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "1            +     .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "2            +     .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "3            +     0  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "4            +     0  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "...        ...   ...                                                ...  \n",
       "3370044      +     .  gene_id \"ENSG00000160299\"; gene_version \"19\"; ...  \n",
       "3370045      +     2  gene_id \"ENSG00000160299\"; gene_version \"19\"; ...  \n",
       "3370046      +     0  gene_id \"ENSG00000160299\"; gene_version \"19\"; ...  \n",
       "3370047      +     .  gene_id \"ENSG00000160299\"; gene_version \"19\"; ...  \n",
       "3370048      +     .  gene_id \"ENSG00000160299\"; gene_version \"19\"; ...  \n",
       "\n",
       "[714883 rows x 9 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ens_gtf = gtf[gtf[\"source\"].str.contains(\"ensembl_havana\")]\n",
    "ens_gtf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "853068e1-8fd3-4af6-b188-bf61103e90ce",
   "metadata": {},
   "source": [
    "# What is the size of ens_df? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "cb6f3b24-2a64-47fb-ad66-5b3bf91e0059",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(714883, 9)"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ens_gtf.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ccb0db5-0f0f-4926-93a9-2a45914b0aaf",
   "metadata": {},
   "source": [
    "# Group ens_df by ‘feature’ column & apply the ‘count’ function to get the number of rows in each feature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "355347af-5def-4dc6-9f00-f4ae845e72a4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feature</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CDS</th>\n",
       "      <td>262675</td>\n",
       "      <td>262675</td>\n",
       "      <td>262675</td>\n",
       "      <td>262675</td>\n",
       "      <td>262675</td>\n",
       "      <td>262675</td>\n",
       "      <td>262675</td>\n",
       "      <td>262675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Selenocysteine</th>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon</th>\n",
       "      <td>286002</td>\n",
       "      <td>286002</td>\n",
       "      <td>286002</td>\n",
       "      <td>286002</td>\n",
       "      <td>286002</td>\n",
       "      <td>286002</td>\n",
       "      <td>286002</td>\n",
       "      <td>286002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>five_prime_utr</th>\n",
       "      <td>38189</td>\n",
       "      <td>38189</td>\n",
       "      <td>38189</td>\n",
       "      <td>38189</td>\n",
       "      <td>38189</td>\n",
       "      <td>38189</td>\n",
       "      <td>38189</td>\n",
       "      <td>38189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <td>19290</td>\n",
       "      <td>19290</td>\n",
       "      <td>19290</td>\n",
       "      <td>19290</td>\n",
       "      <td>19290</td>\n",
       "      <td>19290</td>\n",
       "      <td>19290</td>\n",
       "      <td>19290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>start_codon</th>\n",
       "      <td>25891</td>\n",
       "      <td>25891</td>\n",
       "      <td>25891</td>\n",
       "      <td>25891</td>\n",
       "      <td>25891</td>\n",
       "      <td>25891</td>\n",
       "      <td>25891</td>\n",
       "      <td>25891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>stop_codon</th>\n",
       "      <td>25908</td>\n",
       "      <td>25908</td>\n",
       "      <td>25908</td>\n",
       "      <td>25908</td>\n",
       "      <td>25908</td>\n",
       "      <td>25908</td>\n",
       "      <td>25908</td>\n",
       "      <td>25908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>three_prime_utr</th>\n",
       "      <td>30120</td>\n",
       "      <td>30120</td>\n",
       "      <td>30120</td>\n",
       "      <td>30120</td>\n",
       "      <td>30120</td>\n",
       "      <td>30120</td>\n",
       "      <td>30120</td>\n",
       "      <td>30120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript</th>\n",
       "      <td>26795</td>\n",
       "      <td>26795</td>\n",
       "      <td>26795</td>\n",
       "      <td>26795</td>\n",
       "      <td>26795</td>\n",
       "      <td>26795</td>\n",
       "      <td>26795</td>\n",
       "      <td>26795</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 seqname  source   start     end   score  strand   frame  \\\n",
       "feature                                                                    \n",
       "CDS               262675  262675  262675  262675  262675  262675  262675   \n",
       "Selenocysteine        13      13      13      13      13      13      13   \n",
       "exon              286002  286002  286002  286002  286002  286002  286002   \n",
       "five_prime_utr     38189   38189   38189   38189   38189   38189   38189   \n",
       "gene               19290   19290   19290   19290   19290   19290   19290   \n",
       "start_codon        25891   25891   25891   25891   25891   25891   25891   \n",
       "stop_codon         25908   25908   25908   25908   25908   25908   25908   \n",
       "three_prime_utr    30120   30120   30120   30120   30120   30120   30120   \n",
       "transcript         26795   26795   26795   26795   26795   26795   26795   \n",
       "\n",
       "                 attribute  \n",
       "feature                     \n",
       "CDS                 262675  \n",
       "Selenocysteine          13  \n",
       "exon                286002  \n",
       "five_prime_utr       38189  \n",
       "gene                 19290  \n",
       "start_codon          25891  \n",
       "stop_codon           25908  \n",
       "three_prime_utr      30120  \n",
       "transcript           26795  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(ens_gtf.groupby(\"feature\")\n",
    "   .count() \n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "698facdc-d825-4995-a326-d4afd3e99eda",
   "metadata": {},
   "source": [
    "# Create a new df called ‘genes’ by subsetting all rows with ‘gene’ as a feature "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "d7cf434d-59c6-4025-9fec-64d3d19e0d5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = ens_gtf[ens_gtf[\"feature\"].str.contains(\"gene\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "1ca89f09-2755-477e-8d75-c42de885b182",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>3069168</td>\n",
       "      <td>3438621</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000142611\"; gene_version \"17\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2403964</td>\n",
       "      <td>2413797</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000157911\"; gene_version \"11\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>517</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2228319</td>\n",
       "      <td>2310213</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000157933\"; gene_version \"10\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>580</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>10472288</td>\n",
       "      <td>10630758</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000142655\"; gene_version \"13\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3365880</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>6318378</td>\n",
       "      <td>6360415</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000278903\"; gene_version \"5\"; g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3365953</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>33915534</td>\n",
       "      <td>33977703</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237945\"; gene_version \"9\"; g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3367972</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>9068325</td>\n",
       "      <td>9129773</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000188681\"; gene_version \"13\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3368345</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>14236206</td>\n",
       "      <td>14362754</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000243064\"; gene_version \"10\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3369764</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>46324124</td>\n",
       "      <td>46445769</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160299\"; gene_version \"19\"; ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19290 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname          source feature     start       end score strand  \\\n",
       "0             1  ensembl_havana    gene   1471765   1497848     .      +   \n",
       "132           1  ensembl_havana    gene   3069168   3438621     .      +   \n",
       "361           1  ensembl_havana    gene   2403964   2413797     .      -   \n",
       "517           1  ensembl_havana    gene   2228319   2310213     .      +   \n",
       "580           1  ensembl_havana    gene  10472288  10630758     .      +   \n",
       "...         ...             ...     ...       ...       ...   ...    ...   \n",
       "3365880      21  ensembl_havana    gene   6318378   6360415     .      +   \n",
       "3365953      21  ensembl_havana    gene  33915534  33977703     .      +   \n",
       "3367972      21  ensembl_havana    gene   9068325   9129773     .      -   \n",
       "3368345      21  ensembl_havana    gene  14236206  14362754     .      +   \n",
       "3369764      21  ensembl_havana    gene  46324124  46445769     .      +   \n",
       "\n",
       "        frame                                          attribute  \n",
       "0           .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "132         .  gene_id \"ENSG00000142611\"; gene_version \"17\"; ...  \n",
       "361         .  gene_id \"ENSG00000157911\"; gene_version \"11\"; ...  \n",
       "517         .  gene_id \"ENSG00000157933\"; gene_version \"10\"; ...  \n",
       "580         .  gene_id \"ENSG00000142655\"; gene_version \"13\"; ...  \n",
       "...       ...                                                ...  \n",
       "3365880     .  gene_id \"ENSG00000278903\"; gene_version \"5\"; g...  \n",
       "3365953     .  gene_id \"ENSG00000237945\"; gene_version \"9\"; g...  \n",
       "3367972     .  gene_id \"ENSG00000188681\"; gene_version \"13\"; ...  \n",
       "3368345     .  gene_id \"ENSG00000243064\"; gene_version \"10\"; ...  \n",
       "3369764     .  gene_id \"ENSG00000160299\"; gene_version \"19\"; ...  \n",
       "\n",
       "[19290 rows x 9 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "850347d3-7b47-450f-b5c0-73cbf0d2f942",
   "metadata": {},
   "source": [
    "# Create a new column called ‘gene_length’ and calculate the length of each gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "a28cc35c-50a4-4f25-8e79-5dccb084a96f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/ln/b27g5h050b7c4j9sh03qk90c0000gr/T/ipykernel_96485/4221030734.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  genes[\"gene_length\"] = genes[\"end\"] - genes[\"start\"]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "      <td>26083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>3069168</td>\n",
       "      <td>3438621</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000142611\"; gene_version \"17\"; ...</td>\n",
       "      <td>369453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2403964</td>\n",
       "      <td>2413797</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000157911\"; gene_version \"11\"; ...</td>\n",
       "      <td>9833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>517</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2228319</td>\n",
       "      <td>2310213</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000157933\"; gene_version \"10\"; ...</td>\n",
       "      <td>81894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>580</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>10472288</td>\n",
       "      <td>10630758</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000142655\"; gene_version \"13\"; ...</td>\n",
       "      <td>158470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3365880</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>6318378</td>\n",
       "      <td>6360415</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000278903\"; gene_version \"5\"; g...</td>\n",
       "      <td>42037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3365953</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>33915534</td>\n",
       "      <td>33977703</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237945\"; gene_version \"9\"; g...</td>\n",
       "      <td>62169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3367972</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>9068325</td>\n",
       "      <td>9129773</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000188681\"; gene_version \"13\"; ...</td>\n",
       "      <td>61448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3368345</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>14236206</td>\n",
       "      <td>14362754</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000243064\"; gene_version \"10\"; ...</td>\n",
       "      <td>126548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3369764</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>46324124</td>\n",
       "      <td>46445769</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160299\"; gene_version \"19\"; ...</td>\n",
       "      <td>121645</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19290 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname          source feature     start       end score strand  \\\n",
       "0             1  ensembl_havana    gene   1471765   1497848     .      +   \n",
       "132           1  ensembl_havana    gene   3069168   3438621     .      +   \n",
       "361           1  ensembl_havana    gene   2403964   2413797     .      -   \n",
       "517           1  ensembl_havana    gene   2228319   2310213     .      +   \n",
       "580           1  ensembl_havana    gene  10472288  10630758     .      +   \n",
       "...         ...             ...     ...       ...       ...   ...    ...   \n",
       "3365880      21  ensembl_havana    gene   6318378   6360415     .      +   \n",
       "3365953      21  ensembl_havana    gene  33915534  33977703     .      +   \n",
       "3367972      21  ensembl_havana    gene   9068325   9129773     .      -   \n",
       "3368345      21  ensembl_havana    gene  14236206  14362754     .      +   \n",
       "3369764      21  ensembl_havana    gene  46324124  46445769     .      +   \n",
       "\n",
       "        frame                                          attribute  gene_length  \n",
       "0           .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...        26083  \n",
       "132         .  gene_id \"ENSG00000142611\"; gene_version \"17\"; ...       369453  \n",
       "361         .  gene_id \"ENSG00000157911\"; gene_version \"11\"; ...         9833  \n",
       "517         .  gene_id \"ENSG00000157933\"; gene_version \"10\"; ...        81894  \n",
       "580         .  gene_id \"ENSG00000142655\"; gene_version \"13\"; ...       158470  \n",
       "...       ...                                                ...          ...  \n",
       "3365880     .  gene_id \"ENSG00000278903\"; gene_version \"5\"; g...        42037  \n",
       "3365953     .  gene_id \"ENSG00000237945\"; gene_version \"9\"; g...        62169  \n",
       "3367972     .  gene_id \"ENSG00000188681\"; gene_version \"13\"; ...        61448  \n",
       "3368345     .  gene_id \"ENSG00000243064\"; gene_version \"10\"; ...       126548  \n",
       "3369764     .  gene_id \"ENSG00000160299\"; gene_version \"19\"; ...       121645  \n",
       "\n",
       "[19290 rows x 10 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes[\"gene_length\"] = genes[\"end\"] - genes[\"start\"]\n",
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "e97d732c-2dc5-47a9-be26-5588df933871",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "      <td>26083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>3069168</td>\n",
       "      <td>3438621</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000142611\"; gene_version \"17\"; ...</td>\n",
       "      <td>369453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2403964</td>\n",
       "      <td>2413797</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000157911\"; gene_version \"11\"; ...</td>\n",
       "      <td>9833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>517</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2228319</td>\n",
       "      <td>2310213</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000157933\"; gene_version \"10\"; ...</td>\n",
       "      <td>81894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>580</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>10472288</td>\n",
       "      <td>10630758</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000142655\"; gene_version \"13\"; ...</td>\n",
       "      <td>158470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3365880</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>6318378</td>\n",
       "      <td>6360415</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000278903\"; gene_version \"5\"; g...</td>\n",
       "      <td>42037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3365953</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>33915534</td>\n",
       "      <td>33977703</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237945\"; gene_version \"9\"; g...</td>\n",
       "      <td>62169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3367972</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>9068325</td>\n",
       "      <td>9129773</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000188681\"; gene_version \"13\"; ...</td>\n",
       "      <td>61448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3368345</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>14236206</td>\n",
       "      <td>14362754</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000243064\"; gene_version \"10\"; ...</td>\n",
       "      <td>126548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3369764</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>46324124</td>\n",
       "      <td>46445769</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160299\"; gene_version \"19\"; ...</td>\n",
       "      <td>121645</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19290 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname          source feature     start       end score strand  \\\n",
       "0             1  ensembl_havana    gene   1471765   1497848     .      +   \n",
       "132           1  ensembl_havana    gene   3069168   3438621     .      +   \n",
       "361           1  ensembl_havana    gene   2403964   2413797     .      -   \n",
       "517           1  ensembl_havana    gene   2228319   2310213     .      +   \n",
       "580           1  ensembl_havana    gene  10472288  10630758     .      +   \n",
       "...         ...             ...     ...       ...       ...   ...    ...   \n",
       "3365880      21  ensembl_havana    gene   6318378   6360415     .      +   \n",
       "3365953      21  ensembl_havana    gene  33915534  33977703     .      +   \n",
       "3367972      21  ensembl_havana    gene   9068325   9129773     .      -   \n",
       "3368345      21  ensembl_havana    gene  14236206  14362754     .      +   \n",
       "3369764      21  ensembl_havana    gene  46324124  46445769     .      +   \n",
       "\n",
       "        frame                                          attribute  gene_length  \n",
       "0           .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...        26083  \n",
       "132         .  gene_id \"ENSG00000142611\"; gene_version \"17\"; ...       369453  \n",
       "361         .  gene_id \"ENSG00000157911\"; gene_version \"11\"; ...         9833  \n",
       "517         .  gene_id \"ENSG00000157933\"; gene_version \"10\"; ...        81894  \n",
       "580         .  gene_id \"ENSG00000142655\"; gene_version \"13\"; ...       158470  \n",
       "...       ...                                                ...          ...  \n",
       "3365880     .  gene_id \"ENSG00000278903\"; gene_version \"5\"; g...        42037  \n",
       "3365953     .  gene_id \"ENSG00000237945\"; gene_version \"9\"; g...        62169  \n",
       "3367972     .  gene_id \"ENSG00000188681\"; gene_version \"13\"; ...        61448  \n",
       "3368345     .  gene_id \"ENSG00000243064\"; gene_version \"10\"; ...       126548  \n",
       "3369764     .  gene_id \"ENSG00000160299\"; gene_version \"19\"; ...       121645  \n",
       "\n",
       "[19290 rows x 10 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = genes.assign(gene_length = genes['end'] - genes['start'], \n",
    "                    )\n",
    "genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b83cd827-6f48-4eb9-9886-3e0f6b682474",
   "metadata": {},
   "source": [
    "# Calculate the mean and median of gene lengths "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "2fac1b42-eab9-476b-9387-7a38114d2762",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>1.929000e+04</td>\n",
       "      <td>1.929000e+04</td>\n",
       "      <td>1.929000e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>7.373146e+07</td>\n",
       "      <td>7.379912e+07</td>\n",
       "      <td>6.766304e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>5.582533e+07</td>\n",
       "      <td>5.583178e+07</td>\n",
       "      <td>1.298385e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1.447500e+04</td>\n",
       "      <td>4.687000e+04</td>\n",
       "      <td>6.200000e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>3.110511e+07</td>\n",
       "      <td>3.114020e+07</td>\n",
       "      <td>9.739250e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>5.774250e+07</td>\n",
       "      <td>5.775770e+07</td>\n",
       "      <td>2.736300e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>1.103538e+08</td>\n",
       "      <td>1.104137e+08</td>\n",
       "      <td>7.086475e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>2.489062e+08</td>\n",
       "      <td>2.489199e+08</td>\n",
       "      <td>2.473538e+06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              start           end   gene_length\n",
       "count  1.929000e+04  1.929000e+04  1.929000e+04\n",
       "mean   7.373146e+07  7.379912e+07  6.766304e+04\n",
       "std    5.582533e+07  5.583178e+07  1.298385e+05\n",
       "min    1.447500e+04  4.687000e+04  6.200000e+01\n",
       "25%    3.110511e+07  3.114020e+07  9.739250e+03\n",
       "50%    5.774250e+07  5.775770e+07  2.736300e+04\n",
       "75%    1.103538e+08  1.104137e+08  7.086475e+04\n",
       "max    2.489062e+08  2.489199e+08  2.473538e+06"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c58c3f21-ed5d-4c2e-bd93-870a093399d2",
   "metadata": {},
   "source": [
    "# Find the top 10 longest genes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "8ea3a39a-cabf-4611-8076-9a6df25b7c46",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2679474</th>\n",
       "      <td>16</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>5239802</td>\n",
       "      <td>7713340</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000078328\"; gene_version \"23\"; ...</td>\n",
       "      <td>2473538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1297140</th>\n",
       "      <td>7</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>146116002</td>\n",
       "      <td>148420998</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000174469\"; gene_version \"23\"; ...</td>\n",
       "      <td>2304996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1733179</th>\n",
       "      <td>9</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>8314246</td>\n",
       "      <td>10613002</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000153707\"; gene_version \"19\"; ...</td>\n",
       "      <td>2298756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1514121</th>\n",
       "      <td>X</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>31097677</td>\n",
       "      <td>33339609</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000198947\"; gene_version \"18\"; ...</td>\n",
       "      <td>2241932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1850816</th>\n",
       "      <td>11</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>83455012</td>\n",
       "      <td>85627922</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000150672\"; gene_version \"18\"; ...</td>\n",
       "      <td>2172910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1601567</th>\n",
       "      <td>8</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2935353</td>\n",
       "      <td>4994972</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000183117\"; gene_version \"20\"; ...</td>\n",
       "      <td>2059619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3030998</th>\n",
       "      <td>20</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>13995369</td>\n",
       "      <td>16053197</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000172264\"; gene_version \"18\"; ...</td>\n",
       "      <td>2057828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1104549</th>\n",
       "      <td>6</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>63719980</td>\n",
       "      <td>65707226</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000188107\"; gene_version \"15\"; ...</td>\n",
       "      <td>1987246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2041854</th>\n",
       "      <td>10</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>65912457</td>\n",
       "      <td>67763637</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000183230\"; gene_version \"18\"; ...</td>\n",
       "      <td>1851180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2040306</th>\n",
       "      <td>10</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>53802771</td>\n",
       "      <td>55627942</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000150275\"; gene_version \"20\"; ...</td>\n",
       "      <td>1825171</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname          source feature      start        end score strand  \\\n",
       "2679474      16  ensembl_havana    gene    5239802    7713340     .      +   \n",
       "1297140       7  ensembl_havana    gene  146116002  148420998     .      +   \n",
       "1733179       9  ensembl_havana    gene    8314246   10613002     .      -   \n",
       "1514121       X  ensembl_havana    gene   31097677   33339609     .      -   \n",
       "1850816      11  ensembl_havana    gene   83455012   85627922     .      -   \n",
       "1601567       8  ensembl_havana    gene    2935353    4994972     .      -   \n",
       "3030998      20  ensembl_havana    gene   13995369   16053197     .      +   \n",
       "1104549       6  ensembl_havana    gene   63719980   65707226     .      -   \n",
       "2041854      10  ensembl_havana    gene   65912457   67763637     .      -   \n",
       "2040306      10  ensembl_havana    gene   53802771   55627942     .      -   \n",
       "\n",
       "        frame                                          attribute  gene_length  \n",
       "2679474     .  gene_id \"ENSG00000078328\"; gene_version \"23\"; ...      2473538  \n",
       "1297140     .  gene_id \"ENSG00000174469\"; gene_version \"23\"; ...      2304996  \n",
       "1733179     .  gene_id \"ENSG00000153707\"; gene_version \"19\"; ...      2298756  \n",
       "1514121     .  gene_id \"ENSG00000198947\"; gene_version \"18\"; ...      2241932  \n",
       "1850816     .  gene_id \"ENSG00000150672\"; gene_version \"18\"; ...      2172910  \n",
       "1601567     .  gene_id \"ENSG00000183117\"; gene_version \"20\"; ...      2059619  \n",
       "3030998     .  gene_id \"ENSG00000172264\"; gene_version \"18\"; ...      2057828  \n",
       "1104549     .  gene_id \"ENSG00000188107\"; gene_version \"15\"; ...      1987246  \n",
       "2041854     .  gene_id \"ENSG00000183230\"; gene_version \"18\"; ...      1851180  \n",
       "2040306     .  gene_id \"ENSG00000150275\"; gene_version \"20\"; ...      1825171  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes.sort_values(by=['gene_length'], ascending=False).head(10)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a04b658b-b03a-4472-ab4c-54ee004c3a3e",
   "metadata": {},
   "source": [
    "# Find the lowest 20 shorest genes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "2cb64f15-319e-43a3-99f8-1e0a71e5673f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2913291</th>\n",
       "      <td>17</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>43131208</td>\n",
       "      <td>43131270</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000285057\"; gene_version \"1\"; g...</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2118533</th>\n",
       "      <td>12</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>93083598</td>\n",
       "      <td>93083675</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000256393\"; gene_version \"1\"; g...</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2362811</th>\n",
       "      <td>14</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>22852616</td>\n",
       "      <td>22852728</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000212335\"; gene_version \"1\"; g...</td>\n",
       "      <td>112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>904441</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>90838903</td>\n",
       "      <td>90839037</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000187653\"; gene_version \"11\"; ...</td>\n",
       "      <td>134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1133423</th>\n",
       "      <td>6</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>73373108</td>\n",
       "      <td>73373263</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000235174\"; gene_version \"1\"; g...</td>\n",
       "      <td>155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129959</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>27649419</td>\n",
       "      <td>27649610</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000235912\"; gene_version \"1\"; g...</td>\n",
       "      <td>191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1847776</th>\n",
       "      <td>11</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>82689559</td>\n",
       "      <td>82689768</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000227097\"; gene_version \"5\"; g...</td>\n",
       "      <td>209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2938518</th>\n",
       "      <td>18</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>52409042</td>\n",
       "      <td>52409257</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000280098\"; gene_version \"1\"; g...</td>\n",
       "      <td>215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>675285</th>\n",
       "      <td>3</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>186596976</td>\n",
       "      <td>186597200</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000283390\"; gene_version \"1\"; g...</td>\n",
       "      <td>224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>988497</th>\n",
       "      <td>5</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>62512246</td>\n",
       "      <td>62512482</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000268942\"; gene_version \"2\"; g...</td>\n",
       "      <td>236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1967063</th>\n",
       "      <td>10</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>29697935</td>\n",
       "      <td>29698174</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000225976\"; gene_version \"5\"; g...</td>\n",
       "      <td>239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3255568</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>18992467</td>\n",
       "      <td>18992709</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000185275\"; gene_version \"6\"; g...</td>\n",
       "      <td>242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3363759</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>30718525</td>\n",
       "      <td>30718777</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000231068\"; gene_version \"1\"; g...</td>\n",
       "      <td>252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2274799</th>\n",
       "      <td>12</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>3211663</td>\n",
       "      <td>3211917</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000227081\"; gene_version \"5\"; g...</td>\n",
       "      <td>254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1268412</th>\n",
       "      <td>7</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>129410113</td>\n",
       "      <td>129410370</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000238072\"; gene_version \"1\"; g...</td>\n",
       "      <td>257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105987</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>152622834</td>\n",
       "      <td>152623103</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000185962\"; gene_version \"1\"; g...</td>\n",
       "      <td>269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2594158</th>\n",
       "      <td>16</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>26032539</td>\n",
       "      <td>26032811</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000230330\"; gene_version \"1\"; g...</td>\n",
       "      <td>272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3363751</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>30642864</td>\n",
       "      <td>30643136</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000206104\"; gene_version \"2\"; g...</td>\n",
       "      <td>272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2526556</th>\n",
       "      <td>15</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>29730713</td>\n",
       "      <td>29730985</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000234664\"; gene_version \"1\"; g...</td>\n",
       "      <td>272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3363727</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>30616425</td>\n",
       "      <td>30616699</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000244624\"; gene_version \"3\"; g...</td>\n",
       "      <td>274</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname          source feature      start        end score strand  \\\n",
       "2913291      17  ensembl_havana    gene   43131208   43131270     .      +   \n",
       "2118533      12  ensembl_havana    gene   93083598   93083675     .      -   \n",
       "2362811      14  ensembl_havana    gene   22852616   22852728     .      -   \n",
       "904441        4  ensembl_havana    gene   90838903   90839037     .      -   \n",
       "1133423       6  ensembl_havana    gene   73373108   73373263     .      -   \n",
       "129959        1  ensembl_havana    gene   27649419   27649610     .      +   \n",
       "1847776      11  ensembl_havana    gene   82689559   82689768     .      +   \n",
       "2938518      18  ensembl_havana    gene   52409042   52409257     .      +   \n",
       "675285        3  ensembl_havana    gene  186596976  186597200     .      +   \n",
       "988497        5  ensembl_havana    gene   62512246   62512482     .      -   \n",
       "1967063      10  ensembl_havana    gene   29697935   29698174     .      +   \n",
       "3255568       Y  ensembl_havana    gene   18992467   18992709     .      -   \n",
       "3363759      21  ensembl_havana    gene   30718525   30718777     .      -   \n",
       "2274799      12  ensembl_havana    gene    3211663    3211917     .      -   \n",
       "1268412       7  ensembl_havana    gene  129410113  129410370     .      -   \n",
       "105987        1  ensembl_havana    gene  152622834  152623103     .      -   \n",
       "2594158      16  ensembl_havana    gene   26032539   26032811     .      -   \n",
       "3363751      21  ensembl_havana    gene   30642864   30643136     .      +   \n",
       "2526556      15  ensembl_havana    gene   29730713   29730985     .      +   \n",
       "3363727      21  ensembl_havana    gene   30616425   30616699     .      +   \n",
       "\n",
       "        frame                                          attribute  gene_length  \n",
       "2913291     .  gene_id \"ENSG00000285057\"; gene_version \"1\"; g...           62  \n",
       "2118533     .  gene_id \"ENSG00000256393\"; gene_version \"1\"; g...           77  \n",
       "2362811     .  gene_id \"ENSG00000212335\"; gene_version \"1\"; g...          112  \n",
       "904441      .  gene_id \"ENSG00000187653\"; gene_version \"11\"; ...          134  \n",
       "1133423     .  gene_id \"ENSG00000235174\"; gene_version \"1\"; g...          155  \n",
       "129959      .  gene_id \"ENSG00000235912\"; gene_version \"1\"; g...          191  \n",
       "1847776     .  gene_id \"ENSG00000227097\"; gene_version \"5\"; g...          209  \n",
       "2938518     .  gene_id \"ENSG00000280098\"; gene_version \"1\"; g...          215  \n",
       "675285      .  gene_id \"ENSG00000283390\"; gene_version \"1\"; g...          224  \n",
       "988497      .  gene_id \"ENSG00000268942\"; gene_version \"2\"; g...          236  \n",
       "1967063     .  gene_id \"ENSG00000225976\"; gene_version \"5\"; g...          239  \n",
       "3255568     .  gene_id \"ENSG00000185275\"; gene_version \"6\"; g...          242  \n",
       "3363759     .  gene_id \"ENSG00000231068\"; gene_version \"1\"; g...          252  \n",
       "2274799     .  gene_id \"ENSG00000227081\"; gene_version \"5\"; g...          254  \n",
       "1268412     .  gene_id \"ENSG00000238072\"; gene_version \"1\"; g...          257  \n",
       "105987      .  gene_id \"ENSG00000185962\"; gene_version \"1\"; g...          269  \n",
       "2594158     .  gene_id \"ENSG00000230330\"; gene_version \"1\"; g...          272  \n",
       "3363751     .  gene_id \"ENSG00000206104\"; gene_version \"2\"; g...          272  \n",
       "2526556     .  gene_id \"ENSG00000234664\"; gene_version \"1\"; g...          272  \n",
       "3363727     .  gene_id \"ENSG00000244624\"; gene_version \"3\"; g...          274  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes.sort_values(by=['gene_length']).head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37663efc-e1b9-4cc4-86c8-b5c89c9dbf3d",
   "metadata": {},
   "source": [
    "# Split attributes column on ‘;’ and create new dataframe, merge this onto the ens_gtf dataframe using pandas merge "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "0d885589-be90-4602-b81c-7217f31d9099",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes2=genes[\"attribute\"].str.split(\";\", expand=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "bfc0db26-9af0-49c0-a814-ced562d0719c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene_id \"ENSG00000160072\"</td>\n",
       "      <td>gene_version \"20\"</td>\n",
       "      <td>gene_name \"ATAD3B\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>gene_id \"ENSG00000142611\"</td>\n",
       "      <td>gene_version \"17\"</td>\n",
       "      <td>gene_name \"PRDM16\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>gene_id \"ENSG00000157911\"</td>\n",
       "      <td>gene_version \"11\"</td>\n",
       "      <td>gene_name \"PEX10\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>517</th>\n",
       "      <td>gene_id \"ENSG00000157933\"</td>\n",
       "      <td>gene_version \"10\"</td>\n",
       "      <td>gene_name \"SKI\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>580</th>\n",
       "      <td>gene_id \"ENSG00000142655\"</td>\n",
       "      <td>gene_version \"13\"</td>\n",
       "      <td>gene_name \"PEX14\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3365880</th>\n",
       "      <td>gene_id \"ENSG00000278903\"</td>\n",
       "      <td>gene_version \"5\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"lncRNA\"</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3365953</th>\n",
       "      <td>gene_id \"ENSG00000237945\"</td>\n",
       "      <td>gene_version \"9\"</td>\n",
       "      <td>gene_name \"LINC00649\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"lncRNA\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3367972</th>\n",
       "      <td>gene_id \"ENSG00000188681\"</td>\n",
       "      <td>gene_version \"13\"</td>\n",
       "      <td>gene_name \"TEKT4P2\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"transcribed_unprocessed_pseudog...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3368345</th>\n",
       "      <td>gene_id \"ENSG00000243064\"</td>\n",
       "      <td>gene_version \"10\"</td>\n",
       "      <td>gene_name \"ABCC13\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"transcribed_unprocessed_pseudog...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3369764</th>\n",
       "      <td>gene_id \"ENSG00000160299\"</td>\n",
       "      <td>gene_version \"19\"</td>\n",
       "      <td>gene_name \"PCNT\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19290 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 0                   1  \\\n",
       "0        gene_id \"ENSG00000160072\"   gene_version \"20\"   \n",
       "132      gene_id \"ENSG00000142611\"   gene_version \"17\"   \n",
       "361      gene_id \"ENSG00000157911\"   gene_version \"11\"   \n",
       "517      gene_id \"ENSG00000157933\"   gene_version \"10\"   \n",
       "580      gene_id \"ENSG00000142655\"   gene_version \"13\"   \n",
       "...                            ...                 ...   \n",
       "3365880  gene_id \"ENSG00000278903\"    gene_version \"5\"   \n",
       "3365953  gene_id \"ENSG00000237945\"    gene_version \"9\"   \n",
       "3367972  gene_id \"ENSG00000188681\"   gene_version \"13\"   \n",
       "3368345  gene_id \"ENSG00000243064\"   gene_version \"10\"   \n",
       "3369764  gene_id \"ENSG00000160299\"   gene_version \"19\"   \n",
       "\n",
       "                                     2                              3  \\\n",
       "0                   gene_name \"ATAD3B\"   gene_source \"ensembl_havana\"   \n",
       "132                 gene_name \"PRDM16\"   gene_source \"ensembl_havana\"   \n",
       "361                  gene_name \"PEX10\"   gene_source \"ensembl_havana\"   \n",
       "517                    gene_name \"SKI\"   gene_source \"ensembl_havana\"   \n",
       "580                  gene_name \"PEX14\"   gene_source \"ensembl_havana\"   \n",
       "...                                ...                            ...   \n",
       "3365880   gene_source \"ensembl_havana\"          gene_biotype \"lncRNA\"   \n",
       "3365953          gene_name \"LINC00649\"   gene_source \"ensembl_havana\"   \n",
       "3367972            gene_name \"TEKT4P2\"   gene_source \"ensembl_havana\"   \n",
       "3368345             gene_name \"ABCC13\"   gene_source \"ensembl_havana\"   \n",
       "3369764               gene_name \"PCNT\"   gene_source \"ensembl_havana\"   \n",
       "\n",
       "                                                         4     5  \n",
       "0                            gene_biotype \"protein_coding\"        \n",
       "132                          gene_biotype \"protein_coding\"        \n",
       "361                          gene_biotype \"protein_coding\"        \n",
       "517                          gene_biotype \"protein_coding\"        \n",
       "580                          gene_biotype \"protein_coding\"        \n",
       "...                                                    ...   ...  \n",
       "3365880                                                     None  \n",
       "3365953                              gene_biotype \"lncRNA\"        \n",
       "3367972   gene_biotype \"transcribed_unprocessed_pseudog...        \n",
       "3368345   gene_biotype \"transcribed_unprocessed_pseudog...        \n",
       "3369764                      gene_biotype \"protein_coding\"        \n",
       "\n",
       "[19290 rows x 6 columns]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "469d1f1c-6ea8-40f3-b27a-e9a036542b04",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "      <td>26083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>3069168</td>\n",
       "      <td>3438621</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000142611\"; gene_version \"17\"; ...</td>\n",
       "      <td>369453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2403964</td>\n",
       "      <td>2413797</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000157911\"; gene_version \"11\"; ...</td>\n",
       "      <td>9833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>517</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2228319</td>\n",
       "      <td>2310213</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000157933\"; gene_version \"10\"; ...</td>\n",
       "      <td>81894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>580</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>10472288</td>\n",
       "      <td>10630758</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000142655\"; gene_version \"13\"; ...</td>\n",
       "      <td>158470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3365880</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>6318378</td>\n",
       "      <td>6360415</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000278903\"; gene_version \"5\"; g...</td>\n",
       "      <td>42037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3365953</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>33915534</td>\n",
       "      <td>33977703</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237945\"; gene_version \"9\"; g...</td>\n",
       "      <td>62169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3367972</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>9068325</td>\n",
       "      <td>9129773</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000188681\"; gene_version \"13\"; ...</td>\n",
       "      <td>61448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3368345</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>14236206</td>\n",
       "      <td>14362754</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000243064\"; gene_version \"10\"; ...</td>\n",
       "      <td>126548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3369764</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>46324124</td>\n",
       "      <td>46445769</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160299\"; gene_version \"19\"; ...</td>\n",
       "      <td>121645</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19290 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname          source feature     start       end score strand  \\\n",
       "0             1  ensembl_havana    gene   1471765   1497848     .      +   \n",
       "132           1  ensembl_havana    gene   3069168   3438621     .      +   \n",
       "361           1  ensembl_havana    gene   2403964   2413797     .      -   \n",
       "517           1  ensembl_havana    gene   2228319   2310213     .      +   \n",
       "580           1  ensembl_havana    gene  10472288  10630758     .      +   \n",
       "...         ...             ...     ...       ...       ...   ...    ...   \n",
       "3365880      21  ensembl_havana    gene   6318378   6360415     .      +   \n",
       "3365953      21  ensembl_havana    gene  33915534  33977703     .      +   \n",
       "3367972      21  ensembl_havana    gene   9068325   9129773     .      -   \n",
       "3368345      21  ensembl_havana    gene  14236206  14362754     .      +   \n",
       "3369764      21  ensembl_havana    gene  46324124  46445769     .      +   \n",
       "\n",
       "        frame                                          attribute  gene_length  \n",
       "0           .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...        26083  \n",
       "132         .  gene_id \"ENSG00000142611\"; gene_version \"17\"; ...       369453  \n",
       "361         .  gene_id \"ENSG00000157911\"; gene_version \"11\"; ...         9833  \n",
       "517         .  gene_id \"ENSG00000157933\"; gene_version \"10\"; ...        81894  \n",
       "580         .  gene_id \"ENSG00000142655\"; gene_version \"13\"; ...       158470  \n",
       "...       ...                                                ...          ...  \n",
       "3365880     .  gene_id \"ENSG00000278903\"; gene_version \"5\"; g...        42037  \n",
       "3365953     .  gene_id \"ENSG00000237945\"; gene_version \"9\"; g...        62169  \n",
       "3367972     .  gene_id \"ENSG00000188681\"; gene_version \"13\"; ...        61448  \n",
       "3368345     .  gene_id \"ENSG00000243064\"; gene_version \"10\"; ...       126548  \n",
       "3369764     .  gene_id \"ENSG00000160299\"; gene_version \"19\"; ...       121645  \n",
       "\n",
       "[19290 rows x 10 columns]"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "6b420739-d746-4a16-b944-c7fe73a83af1",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes3 = genes.join(genes2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "bb8b19f5-886c-47cf-88fb-68af1cde53bd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_length</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "      <td>26083</td>\n",
       "      <td>gene_id \"ENSG00000160072\"</td>\n",
       "      <td>gene_version \"20\"</td>\n",
       "      <td>gene_name \"ATAD3B\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>3069168</td>\n",
       "      <td>3438621</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000142611\"; gene_version \"17\"; ...</td>\n",
       "      <td>369453</td>\n",
       "      <td>gene_id \"ENSG00000142611\"</td>\n",
       "      <td>gene_version \"17\"</td>\n",
       "      <td>gene_name \"PRDM16\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2403964</td>\n",
       "      <td>2413797</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000157911\"; gene_version \"11\"; ...</td>\n",
       "      <td>9833</td>\n",
       "      <td>gene_id \"ENSG00000157911\"</td>\n",
       "      <td>gene_version \"11\"</td>\n",
       "      <td>gene_name \"PEX10\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>517</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2228319</td>\n",
       "      <td>2310213</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000157933\"; gene_version \"10\"; ...</td>\n",
       "      <td>81894</td>\n",
       "      <td>gene_id \"ENSG00000157933\"</td>\n",
       "      <td>gene_version \"10\"</td>\n",
       "      <td>gene_name \"SKI\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>580</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>10472288</td>\n",
       "      <td>10630758</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000142655\"; gene_version \"13\"; ...</td>\n",
       "      <td>158470</td>\n",
       "      <td>gene_id \"ENSG00000142655\"</td>\n",
       "      <td>gene_version \"13\"</td>\n",
       "      <td>gene_name \"PEX14\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3365880</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>6318378</td>\n",
       "      <td>6360415</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000278903\"; gene_version \"5\"; g...</td>\n",
       "      <td>42037</td>\n",
       "      <td>gene_id \"ENSG00000278903\"</td>\n",
       "      <td>gene_version \"5\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"lncRNA\"</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3365953</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>33915534</td>\n",
       "      <td>33977703</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237945\"; gene_version \"9\"; g...</td>\n",
       "      <td>62169</td>\n",
       "      <td>gene_id \"ENSG00000237945\"</td>\n",
       "      <td>gene_version \"9\"</td>\n",
       "      <td>gene_name \"LINC00649\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"lncRNA\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3367972</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>9068325</td>\n",
       "      <td>9129773</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000188681\"; gene_version \"13\"; ...</td>\n",
       "      <td>61448</td>\n",
       "      <td>gene_id \"ENSG00000188681\"</td>\n",
       "      <td>gene_version \"13\"</td>\n",
       "      <td>gene_name \"TEKT4P2\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"transcribed_unprocessed_pseudog...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3368345</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>14236206</td>\n",
       "      <td>14362754</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000243064\"; gene_version \"10\"; ...</td>\n",
       "      <td>126548</td>\n",
       "      <td>gene_id \"ENSG00000243064\"</td>\n",
       "      <td>gene_version \"10\"</td>\n",
       "      <td>gene_name \"ABCC13\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"transcribed_unprocessed_pseudog...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3369764</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>46324124</td>\n",
       "      <td>46445769</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160299\"; gene_version \"19\"; ...</td>\n",
       "      <td>121645</td>\n",
       "      <td>gene_id \"ENSG00000160299\"</td>\n",
       "      <td>gene_version \"19\"</td>\n",
       "      <td>gene_name \"PCNT\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19290 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname          source feature     start       end score strand  \\\n",
       "0             1  ensembl_havana    gene   1471765   1497848     .      +   \n",
       "132           1  ensembl_havana    gene   3069168   3438621     .      +   \n",
       "361           1  ensembl_havana    gene   2403964   2413797     .      -   \n",
       "517           1  ensembl_havana    gene   2228319   2310213     .      +   \n",
       "580           1  ensembl_havana    gene  10472288  10630758     .      +   \n",
       "...         ...             ...     ...       ...       ...   ...    ...   \n",
       "3365880      21  ensembl_havana    gene   6318378   6360415     .      +   \n",
       "3365953      21  ensembl_havana    gene  33915534  33977703     .      +   \n",
       "3367972      21  ensembl_havana    gene   9068325   9129773     .      -   \n",
       "3368345      21  ensembl_havana    gene  14236206  14362754     .      +   \n",
       "3369764      21  ensembl_havana    gene  46324124  46445769     .      +   \n",
       "\n",
       "        frame                                          attribute  gene_length  \\\n",
       "0           .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...        26083   \n",
       "132         .  gene_id \"ENSG00000142611\"; gene_version \"17\"; ...       369453   \n",
       "361         .  gene_id \"ENSG00000157911\"; gene_version \"11\"; ...         9833   \n",
       "517         .  gene_id \"ENSG00000157933\"; gene_version \"10\"; ...        81894   \n",
       "580         .  gene_id \"ENSG00000142655\"; gene_version \"13\"; ...       158470   \n",
       "...       ...                                                ...          ...   \n",
       "3365880     .  gene_id \"ENSG00000278903\"; gene_version \"5\"; g...        42037   \n",
       "3365953     .  gene_id \"ENSG00000237945\"; gene_version \"9\"; g...        62169   \n",
       "3367972     .  gene_id \"ENSG00000188681\"; gene_version \"13\"; ...        61448   \n",
       "3368345     .  gene_id \"ENSG00000243064\"; gene_version \"10\"; ...       126548   \n",
       "3369764     .  gene_id \"ENSG00000160299\"; gene_version \"19\"; ...       121645   \n",
       "\n",
       "                                 0                   1  \\\n",
       "0        gene_id \"ENSG00000160072\"   gene_version \"20\"   \n",
       "132      gene_id \"ENSG00000142611\"   gene_version \"17\"   \n",
       "361      gene_id \"ENSG00000157911\"   gene_version \"11\"   \n",
       "517      gene_id \"ENSG00000157933\"   gene_version \"10\"   \n",
       "580      gene_id \"ENSG00000142655\"   gene_version \"13\"   \n",
       "...                            ...                 ...   \n",
       "3365880  gene_id \"ENSG00000278903\"    gene_version \"5\"   \n",
       "3365953  gene_id \"ENSG00000237945\"    gene_version \"9\"   \n",
       "3367972  gene_id \"ENSG00000188681\"   gene_version \"13\"   \n",
       "3368345  gene_id \"ENSG00000243064\"   gene_version \"10\"   \n",
       "3369764  gene_id \"ENSG00000160299\"   gene_version \"19\"   \n",
       "\n",
       "                                     2                              3  \\\n",
       "0                   gene_name \"ATAD3B\"   gene_source \"ensembl_havana\"   \n",
       "132                 gene_name \"PRDM16\"   gene_source \"ensembl_havana\"   \n",
       "361                  gene_name \"PEX10\"   gene_source \"ensembl_havana\"   \n",
       "517                    gene_name \"SKI\"   gene_source \"ensembl_havana\"   \n",
       "580                  gene_name \"PEX14\"   gene_source \"ensembl_havana\"   \n",
       "...                                ...                            ...   \n",
       "3365880   gene_source \"ensembl_havana\"          gene_biotype \"lncRNA\"   \n",
       "3365953          gene_name \"LINC00649\"   gene_source \"ensembl_havana\"   \n",
       "3367972            gene_name \"TEKT4P2\"   gene_source \"ensembl_havana\"   \n",
       "3368345             gene_name \"ABCC13\"   gene_source \"ensembl_havana\"   \n",
       "3369764               gene_name \"PCNT\"   gene_source \"ensembl_havana\"   \n",
       "\n",
       "                                                         4     5  \n",
       "0                            gene_biotype \"protein_coding\"        \n",
       "132                          gene_biotype \"protein_coding\"        \n",
       "361                          gene_biotype \"protein_coding\"        \n",
       "517                          gene_biotype \"protein_coding\"        \n",
       "580                          gene_biotype \"protein_coding\"        \n",
       "...                                                    ...   ...  \n",
       "3365880                                                     None  \n",
       "3365953                              gene_biotype \"lncRNA\"        \n",
       "3367972   gene_biotype \"transcribed_unprocessed_pseudog...        \n",
       "3368345   gene_biotype \"transcribed_unprocessed_pseudog...        \n",
       "3369764                      gene_biotype \"protein_coding\"        \n",
       "\n",
       "[19290 rows x 16 columns]"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "53753311-c9a3-477d-a4f4-6ec911c4ac5e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(19290, 16)"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes3.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "9a47a9bd-650b-4273-8950-836185bf1c40",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(19290, 6)"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes2.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "a4ad3b72-896e-4a94-8322-4cca77007b2b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(19290, 10)"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8ef9be7-1905-491b-bcec-b7e4d0227de1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
